[M]33.cpp

```diff
@@ -0,0 +1,36 @@
+// Problem: 33. Search in Rotated Sorted Array
+// Link: https://leetcode.com/problems/search-in-rotated-sorted-array/
+
+class Solution {
+public:
+    int find(const vector<int> &nums, int l, int r, int target) {
+        if (l == r) {
+            return nums[l] == target? l : -1;
+        }
+        else if (l > r) {
+            return -1;
+        }
+        
+        int mid = l + (r - l) / 2;
+        int res;
+        if (nums[mid] == target) {
+            return mid;
+        }
+        else if (nums[mid] < target) {
+            res = find(nums, mid + 1, r, target);
```

Comment:
Code thế này liệu độ phức tạp có còn là O(log(N)) nữa ko? vì chú gọi 2 lần nữa. Dòng 26,27 cũng thế.

Reply:
Độ phức tạp trong trường hợp này theo anh là khá khó tính toán, nó có thể không phải là O(log(n)) nhưng sẽ không đến O(n). Anh phân tích thế này:

- Phép tính tích cực trong trường hợp này là int mid = l + (r - l) / 2;
- Dãy ban đầu sau khi được rotate, nếu dùng mid chia ra 2 đoạn, thì trong đó 1 đoạn đơn điệu tăng như bình thường, và 1 đoạn gồm 2 đoạn đơn điệu tăng nhỏ hơn.
- Ví dụ: 5 6 7 8 9 10 (mid = 11) 12 13 1 2 3 4 
=> Bên trái đoạn bình thường. Bên phải gọi là abnormal gồm 2 đoạn nhỏ hơn cũng đơn điệu tăng.
- Cách làm của anh là sẽ search đoạn bình thường trước, sau đó sẽ search đoạn abnormal. Cứ đệ quy như vậy với các đoạn nhỏ hơn.
- Trường hợp search đoạn bình thường thì độ phức tạp sẽ là O(log(N)) như bình thường.
- Trường hợp search đoạn abnomal sẽ là O(log(n/2))
- Vì thế, theo anh nghĩ độ phức tạp có thể là: O(log(n)) + O(log(n/2)) + O(log(n/4)) ...

Reply:
Chú thử refactor lại cho nó thành exactly O(log(N)) xem.